Bug: Bad file descriptor (os error 9) during data replication

### What happened?

Sometimes I get an error in the logs after restarting an instance with provisioned replication:

```
2024-11-12 00:07:51.322 (87136) [ERROR] -- replication/replication_sender.rs:129 Failed to remove transaction: ReductError { status: InternalServerError, message: "Bad file descriptor (os error 9)" }
```

Looks like a problem with the cache for file descriptors. 

### Version

v1.12.3

### OS

Linux

### Snippet to repeat the bug

_No response_

### Logs

```Shell
2024-11-12 00:07:49.815 (54944) [INFO] -- main.rs:23 ReductStore 1.12.3 [e27a887 at 2024-11-11T19:41:00Z]
2024-11-12 00:07:49.835 (54944) [INFO] -- main.rs:38 Configuration: 
 	RS_LOG_LEVEL = INFO 
	RS_HOST = 0.0.0.0 (default)
	RS_PORT = 8383 (default)
	RS_API_BASE_PATH = / (default)
	RS_DATA_PATH = /data (default)
	RS_API_TOKEN = *********** 
	RS_BUCKET_2_NAME = repl 
	RS_BUCKET_3_NAME = stress_test 
	RS_BUCKET_2_QUOTA_TYPE = FIFO 
	RS_BUCKET_2_QUOTA_SIZE = 5.0 GB 
	RS_BUCKET_3_QUOTA_TYPE = FIFO 
	RS_BUCKET_3_QUOTA_SIZE = 10.0 GB 
	RS_TOKEN_2_NAME = repl-rw 
	RS_TOKEN_3_NAME = stress_test-rw 
	RS_TOKEN_2_VALUE = ********************************* 
	RS_TOKEN_3_VALUE = *********************************** 
	RS_TOKEN_2_READ = repl 
	RS_TOKEN_2_WRITE = repl 
	RS_TOKEN_3_READ = stress_test 
	RS_TOKEN_3_WRITE = stress_test 
	RS_TOKEN_3_FULL_ACCESS = false 
	RS_REPLICATION_1_NAME = stress_test 
	RS_REPLICATION_1_SRC_BUCKET = stress_test 
	RS_REPLICATION_1_DST_BUCKET = repl 
	RS_REPLICATION_1_DST_HOST = http://reductstore:8383 
	RS_REPLICATION_1_DST_TOKEN = ********************************* 
	RS_REPLICATION_1_INCLUDE_size = medium 

2024-11-12 00:07:49.875 (41568) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/data/stress_test/entry-1"
2024-11-12 00:07:49.876 (47264) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/data/stress_test/entry-2"
2024-11-12 00:07:49.973 (35872) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/data/repl/entry-1"
2024-11-12 00:07:49.986 (47264) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/data/repl/entry-2"
2024-11-12 00:07:50.086 (54944) [INFO] -- storage/storage.rs:67 Load 2 buckets
2024-11-12 00:07:50.086 (54944) [INFO] -- reductstore/src/cfg.rs:146 Provisioned bucket 'stress_test' with: BucketSettings { quota_type: Some(FIFO), quota_size: Some(10000000000), max_block_size: Some(64000000), max_block_records: Some(1024) }
2024-11-12 00:07:50.086 (54944) [INFO] -- reductstore/src/cfg.rs:146 Provisioned bucket 'repl' with: BucketSettings { quota_type: Some(FIFO), quota_size: Some(5000000000), max_block_size: Some(64000000), max_block_records: Some(1024) }
2024-11-12 00:07:50.096 (54944) [INFO] -- reductstore/src/cfg.rs:112 Provisioned token 'stress_test-rw' with Some(Permissions { full_access: false, read: ["stress_test"], write: ["stress_test"] })
2024-11-12 00:07:50.096 (54944) [INFO] -- reductstore/src/cfg.rs:112 Provisioned token 'repl-rw' with Some(Permissions { full_access: false, read: ["repl"], write: ["repl"] })
2024-11-12 00:07:50.096 (54944) [INFO] -- cfg.rs:372 Load Web Console
2024-11-12 00:07:50.709 (54944) [INFO] -- reductstore/src/cfg.rs:176 Provisioned replication 'stress_test' with ReplicationSettings { src_bucket: "stress_test", dst_bucket: "repl", dst_host: "http://reductstore:8383/", dst_token: "***", entries: [], include: {"size": "medium"}, exclude: {}, each_n: None, each_s: None }
2024-11-12 00:07:50.709 (54944) [INFO] -- main.rs:48 License: BUSL-1.1 [https://github.com/reductstore/reductstore/blob/v1.12.3/LICENSE]
2024-11-12 00:07:50.709 (54944) [INFO] -- main.rs:60 Run HTTP reductstore on http://0.0.0.0:8383/
2024-11-12 00:07:51.322 (87136) [ERROR] -- replication/replication_sender.rs:129 Failed to remove transaction: ReductError { status: InternalServerError, message: "Bad file descriptor (os error 9)" }
2024-11-12 00:07:52.106 (70048) [ERROR] -- replication/replication_sender.rs:129 Failed to remove transaction: ReductError { status: InternalServerError, message: "Bad file descriptor (os error 9)" }
2024-11-12 00:07:52.217 (87136) [ERROR] -- replication/replication_sender.rs:129 Failed to remove transaction: ReductError { status: InternalServerError, message: "Bad file descriptor (os error 9)" }
2024-11-12 00:07:52.378 (70048) [ERROR] -- replication/replication_sender.rs:129 Failed to remove transaction: ReductError { status: InternalServerError, message: "Bad file descriptor (os error 9)" }
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+
+- RS-539: Fix transaction log file descriptor, [PR-643](https://github.com/reductstore/reductstore/pull/643)
+
 ## [1.12.3] - 2024-10-26
 
 ### Fixed

--- a/reductstore/src/replication/transaction_log.rs
+++ b/reductstore/src/replication/transaction_log.rs
@@ -227,7 +227,7 @@ impl TransactionLog {
 
         {
             let file = FILE_CACHE
-                .read(&self.file_path, SeekFrom::Start(8))?
+                .write_or_create(&self.file_path, SeekFrom::Start(8))?
                 .upgrade()?;
             let mut file = file.write()?;
             file.write_all(&self.read_pos.to_be_bytes())?;
@@ -363,13 +363,14 @@ mod tests {
 
     #[rstest]
     fn test_recovery_empty_cache(path: PathBuf) {
-        TransactionLog::try_load_or_create(path.clone(), 3).unwrap();
-        FILE_CACHE.discard_recursive(&path).unwrap(); // discard the cache to simulate restart
-
-        let mut transaction_log = TransactionLog::try_load_or_create(path, 3).unwrap();
+        let mut transaction_log = TransactionLog::try_load_or_create(path.clone(), 3).unwrap();
         transaction_log
             .push_back(Transaction::WriteRecord(1))
             .unwrap();
+
+        FILE_CACHE.discard_recursive(&path).unwrap(); // discard the cache to simulate restart
+
+        let mut transaction_log = TransactionLog::try_load_or_create(path, 3).unwrap();
 
         // check if the transaction log is still working after cache discard
         assert_eq!(

